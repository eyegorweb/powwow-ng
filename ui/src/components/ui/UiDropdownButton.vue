<template>
  <div class="position-relative p-0">
    <slot name="trigger" static-class="dropdown-toggle" :is-open="isOpen" :toggle="toggle">
      <UiButton
        variant="outline-gray"
        class="dropdown-toggle"
        :class="isOpen && 'text-secondary'"
        @click.stop="toggle"
      >
        <slot name="button" />
      </UiButton>
    </slot>

    <transition name="fade">
      <div
        class="dropdown-menu shadow rounded-1 border-0 show p-0"
        :style="menuStyle"
        v-if="isOpen"
        v-clickaway="hide"
      >
        <slot :hide="hide" />
      </div>
    </transition>

    <slot name="append" />
  </div>
</template>

<script>
import { clickaway } from '@/directives/clickaway';
import UiButton from '@/components/ui/Button';

export default {
  props: {
    menuStyle: {
      type: Object,
    },
  },
  data() {
    return {
      isOpen: false,
    };
  },

  methods: {
    toggle() {
      this.isOpen = !this.isOpen;
    },
    hide() {
      this.isOpen = false;
    },
  },

  directives: { clickaway },
  components: { UiButton },
};
</script>

<style scoped></style>
