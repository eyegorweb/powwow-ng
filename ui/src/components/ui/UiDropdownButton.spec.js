import { mount } from '@vue/test-utils';
import UiDropdownButton from './UiDropdownButton.vue';

describe('UiDropdownButton.vue', () => {
  it('toggles the dropdown', async () => {
    const wrapper = mount(UiDropdownButton);
    expect(wrapper.find('.dropdown-menu').exists()).toBe(false);
    await wrapper.find('.dropdown-toggle').trigger('click');
    expect(wrapper.find('.dropdown-menu').exists()).toBe(true);
    await wrapper.find('.dropdown-toggle').trigger('click');
    expect(wrapper.find('.dropdown-menu').exists()).toBe(false);
  });

  it('can be hidden', async () => {
    const wrapper = mount(UiDropdownButton);
    await wrapper.find('.dropdown-toggle').trigger('click');
    expect(wrapper.find('.dropdown-menu').exists()).toBe(true);
    await wrapper.vm.hide();
    expect(wrapper.find('.dropdown-menu').exists()).toBe(false);
  });

  it('can customise the button', async () => {
    const wrapper = mount(UiDropdownButton, {
      scopedSlots: {
        trigger: `<div :class="props.staticClass" @click.stop="props.toggle">{{ props.isOpen }}</div>`,
      },
    });
    expect(wrapper.find('.dropdown-menu').exists()).toBe(false);
    expect(wrapper.find('.dropdown-toggle').text()).toBe('false');
    await wrapper.find('.dropdown-toggle').trigger('click');
    expect(wrapper.find('.dropdown-menu').exists()).toBe(true);
    expect(wrapper.find('.dropdown-toggle').text()).toBe('true');
  });

  it('can append content', () => {
    const wrapper = mount(UiDropdownButton, {
      scopedSlots: {
        append: `<div>appended</div>`,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
