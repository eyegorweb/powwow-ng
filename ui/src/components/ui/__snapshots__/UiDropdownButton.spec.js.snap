// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiDropdownButton.vue can append content 1`] = `
<div class="position-relative p-0"><button type="button" class="dropdown-toggle btn btn-outline-gray"></button>
  <transition-stub name="fade">
    <!---->
  </transition-stub>
  <div>appended</div>
</div>
`;
